pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	if err = pc.cniServer.GetPodConfigurator().ConfigureSriovSecondaryInterface(
+		containerInfo.PodName,
+		containerInfo.PodNameSpace,
+		containerInfo.ContainerID,
+		containerInfo.ContainerNetNS,
+		netinfo.InterfaceName,
+		containerInfo.MTU,
+		podSriovVFDeviceID[0],
+		result,
+	); err != nil {
+		return fmt.Errorf("SRIOV Interface creation failed: %v", err)
+	}
+	return nil
+}
+
+// Configure Secondary Network Interface.
+func (pc *PodController) configureSecondaryInterface(pod *corev1.Pod, netinfo *SecondaryNetworkObject, podCNIInfo *cnipodcache.CNIConfigInfo, cniconfig []byte) error {
+	var err error
+	var ipamerr error
+	var ipamResult *current.Result
+	var cmdArgs *invoke.Args
+
+	// Generate and assign new interface name, If secondary interface name was not provided in Pod annotation.
+	if len(netinfo.InterfaceName) == 0 {
+		netinfo.InterfaceName = generatePodSecondaryIfaceName(podCNIInfo)
+	}
+	if netinfo.InterfaceType == sriovInterfaceType {
+		cmdArgs = &invoke.Args{Command: string("ADD"), ContainerID: podCNIInfo.ContainerID,
+			NetNS: podCNIInfo.ContainerNetNS, IfName: netinfo.InterfaceName,
+			Path: cniPath}
+
+		ipamResult, err = ipam.GetIPAMSubnetAddress(cniconfig, cmdArgs)
+		if err != nil {
+			return errors.New("secondary network IPAM failed")
+		}
+
+		result := &current.Result{CNIVersion: podCNIInfo.CNIVersion}
+		result.IPs = ipamResult.IPs
+		result.Routes = ipamResult.Routes
+		// Set result.Interface to container interface.
+		for _, ip := range result.IPs {
+			ip.Interface = current.Int(1)
+		}
+		// Configure SRIOV as a secondary network interface
+		if err = pc.configureSriovAsSecondaryInterface(pod, netinfo, podCNIInfo, result); err != nil {
+			// SRIOV interface creation failed. Free allocated IP address
+			if ipamerr = ipam.DelIPAMSubnetAddress(cniconfig, cmdArgs); ipamerr != nil {
+				klog.ErrorS(err, "IPAM de-allocation failed: ", ipamerr)
+			}
+			return err
+		}
+		// Update Pod CNI cache with the network config which was successfully configured.
+		if podCNIInfo.NetworkConfig == nil {
+			podCNIInfo.NetworkConfig = make(map[string][]byte)
+		}
+		podCNIInfo.NetworkConfig[netinfo.InterfaceName] = cniconfig
+	} else {
+		klog.ErrorS(err, "InterfaceType not supported for Pod ", klog.KObj(pod))
+	}
+	return nil
+}
+
+func (pc *PodController) configureSecondaryNetwork(pod *corev1.Pod, networklist []*SecondaryNetworkObject, podCNIInfo *cnipodcache.CNIConfigInfo) error {
+
+	for _, netinfo := range networklist {
+		klog.InfoS("Secondary Network Information:", netinfo)
+		if len(netinfo.NetworkName) > 0 {
+			netDefCRD, err := pc.netAttachDefClient.NetworkAttachmentDefinitions(pod.Namespace).Get(context.TODO(), netinfo.NetworkName, metav1.GetOptions{})
+			if err != nil {
+				// NetworkAttachmentDefinition not found at this time. Return error to re-queue and re-try.
+				return fmt.Errorf("NetworkAttachmentDefinitions.Get failed %v", err)
```

Comment:
we usually precede the `%v` formatter (or `%w`) with a colon and a space when we wrap errors (see rest of codebase)